### PR TITLE
Better error handling

### PR DESCRIPTION
**Pull request by mkangia**
https://manage.dimagi.com/default.asp?276927

Show error message to user 
```
Looks like you are assigning a location name/code to a different location in the same request. 
Please do this in two separate updates by using a temporary name to free up the name/code to be re-assigned.
```

if they are trying to swap names/codes for two locations or trying to use name/code for a location to another

Doing such an update in a single request results in IntegrityError at database due to unique constraint while the updates are being done sequentially.

fyi @emord @esoergel

### CODE REVIEW

File: corehq/apps/locations/views.py
Comment (esoergel) [inline corehq/apps/locations/views.py]: I think we should be able to work around this in some way - would wrapping all the saves in a single transaction allow this to go through as-is?

Comment (mkangia) [reply corehq/apps/locations/views.py]: @esoergel i don't think that would help since the save would still happen one by one and the unique clause would fail?

Comment (esoergel) [reply corehq/apps/locations/views.py]: Dang, I was hoping postgres would defer unique checking until the transaction is finished, but that doesn't appear to be the case (just tried it out in a shell session).

Another option is to assign one of the location types a temporary name (say, a UUID), then make a second pass and save it again with the real name.  It's an extra round-trip to the db, but we expect there to be a very small number of location types on any given domain, and it beats making the user do that manually.

Comment (esoergel) [reply corehq/apps/locations/views.py]: Huh, okay.  I do think that would be better behavior, but I haven't tried it to see how complex that is to implement.  I'll leave it up to your judgement on whether it's worth it.

**Review by esoergel — APPROVED** [review]

File: corehq/apps/locations/views.py
Comment (snopoke) [reply corehq/apps/locations/views.py]: I had similar thoughts as @esoergel but I guess if it does introduce a lot of complexity then fine to leave as is for now. If this comes up again we can do it then.

**Comment by mkangia** [comment]
@esoergel 
any suggestions on the stickler error i am getting here. Can it be ignored?

File: corehq/apps/locations/views.py
Comment (mkangia) [reply corehq/apps/locations/views.py]: Thank you.

**Comment by esoergel** [comment]
@mkangia 
> W1656 dict.values referenced when not iterating

That's a python 2-to-3 issue - in Python 3, `.values()` returns an iterator, which can break some things.  The fix is to immediately wrap the call with `list`

```python
names = list(payload_loc_type_name_by_pk.values())
```

**Comment by mkangia** [comment]
Cool, thanks @esoergel

**Comment by mkangia** [comment]
Could not find a setup to test views so added a new one and added some tests.
@esoergel

**Review by esoergel — APPROVED** [review]

**Comment by mkangia** [comment]
@dimagi/product this should be happening very often but just fyi in case you have any feedback/suggestion.

**Comment by mkangia** [comment]
Offlined with @djmore9 and this is okay.

**Comment by snopoke** [comment]
:+1: still has the 'do not merge' label

**Comment by mkangia** [comment]
@snopoke 
removed the label and merged.
Thank you!

**Pull request merged**
